### DISCUSSION TITLE
Markdown syntax for coloured hyperlinks?

### DISCUSSION THREAD

**Discussion by no-vici** [Omikhleia/markdown.sile#66] [Q&A] [answered]
Hi again Omikhleia,

I would like to include coloured hyperlinks in my `.md` source file using Markdown syntax. Does `markdown.sile` have this feature? I tried using `<span style="color:red"> *some emphasized markdown text*</span>`, but it did not work. I also skimmed through the manual but did not find a solution. While I can inject SILE code into my `.md` to achieve this, it would be preferable to use what `markdown.sile` might have already offered. This would be cleaner and more seamless conversion of the source file into other formats using Pandoc.

**Comment by Omikhleia** [comment]
Interesting. 
I assume you mean links with a text representation, not just raw URLs?

There could be several solutions, possibly...

Here is one of them:

If using `resilient.sile` and the **resilient.book** class, it is achievable, although a bit contrived currently.

Say you have a link like this: `[Markdown](https://en.wikipedia.org/wiki/Markdown)`
You could rewrite like this `[[Markdown](https://en.wikipedia.org/wiki/Markdown)]{custom-style=MyLink}` (that is, wrapping it in a Pandoc-like span element, with a custom style attribute).
Then in your style file, you could then add:

```
MyLink:
  style:
    color: "blue"
```

This strategy has the disadvantage of having you modifying all your links in your source (which you might have been ready to do, with HTML `<span>` elements[^1], but still...).
It has the advantage of using a Pandoc-compatible syntax (avoiding HTML stuff or explicit SILE code), and you could have more than one custom style, and change colors in one place only)[^2].

I can try to work on a fix that would avoid the extra Pandoc span, i.e. using `[Markdown](https://en.wikipedia.org/wiki/Markdown){custom-style=MyLink}` with the attribute directly on the link element. This should have worked, but I overlooked it apparently. I opened  #67 to track it.

[^1]: Markdown.sile doesn't support them currently (#29). That would require some work, and I am not sure I'll consider it at all, as Djot also moves away from mixing lightweight mark-up with raw HTML.

[^2]: At some point I'll want the styling paradigm to support underlining, too -- but not until mutli-line links are supported in SILE. For now it would be doable, but not so nicely: `[[Markdown](https://en.wikipedia.org/wiki/Markdown)]{.underline custom-style=MyLink}`

**Reply by Omikhleia** [reply]
> I opened https://github.com/Omikhleia/markdown.sile/issues/67 to track it.

If you are not in a hurry, I might push the fix later (I must have dinner, etc. lol... and I want to test it well, but it's a quick thing normally). That would avoid the double-span effect, at least... --> **EDIT**: Done :heavy_check_mark:

**Reply by no-vici** [reply]
> > I opened #67 to track it.
> 
> If you are not in a hurry, I might push the fix later (I must have dinner, etc. lol... and I want to test it well, but it's a quick thing normally). That would avoid the double-span effect, at least... --> **EDIT**: Done ✔️

Great thanks for your quick reply and valuable time.

I am learning SILE for fun so obviously not in a hurry at all ;). I played with what I wanted to have a bit late last night, and asked the question a few mins before bedtime. Of course, your dinner and well-being is more important than helping me though.

I will give your suggestions a go now and let you know the result soon.

PS. Please tell me how would I buy you a few beers then by dropping me a few lines at bluehatvietnam [at] gmail.

**Reply by no-vici** [reply]
> `[[Markdown](https://en.wikipedia.org/wiki/Markdown)]{custom-style=MyLink}`

This yields no colour. Modifying the source is quickly achieved with a Neovim's macro or a lua snippets and I don't mind either of them.

> `[[Markdown](https://en.wikipedia.org/wiki/Markdown)]{.underline custom-style=MyLink}`

The underline appears but still no colour.

I can put the raw URL's as hyperlinks so the reader can copy and paste or when they hover the cursor over the text they can click on it.

It would be nice to see the colour for hyperlinks when covert MD to PDF by SILE if the implementation is not that hard, if it is, some manual labour would do the trick though ;)

Thank you again for looking into it.

Cheers.

**Reply by Omikhleia** [reply, marked as answer]
> The underline appears but still no colour.

Hm. If the underline works... It might mean it doesn't find the MyLink style. I should add a debug option there to understand. It might be something tricky, or my setup is different (that's the problem with so many packages, I don't manage versions very properly and I'm running the very latest versions of resilient.sile and markdown.sile... perhaps I fixed something in one dependency without even noticing it ^^

I was thinking, maybe you could have your book's sources on a _private_ GitHub repository and add me as contributor there, so I can try to build it? I can sign you a non-disclosure agreement, if you want ;-)

One thing you might try if underlining works and you have a SIL preamble is to have some `\define[command=Blue]{\color[color=blue]{\process}}` and use that as custom style: `[[Markdown](https://....)]{.underline custom-style=Blue}` -- To explain: custom styles in the current implementation are supposed to use (resilient) styles if used with them and one by the said name exists; or SILE commands as a fallback if one by the same name exists; or do nothing.

**Reply by no-vici** [reply]
> `[[Markdown](https://....)]{.underline custom-style=Blue}`

This trick obviously works as I thought before compiling the source, and it's a nice trick too, so thank you.

> I was thinking, maybe you could have your book's sources on a private GitHub repository and add me as contributor there, so I can try to build it? I can sign you a non-disclosure agreement, if you want ;-)

Thank you for this very generous offer, this would be a very grateful help. I will have to learn how to setup a repo hosted on GitHub, I have never done this, so it will take me some time to learn ;).

All the book projects that I've been playing with are history books about The Vietnam War, which will then be shared publicly to tell young Vietnamese's about what called "relatively true history" of the war that their grandpas blindly fought for. The history that the totalitarian regime in Vietnam always tries to hide from its people. When you looks at the CCP, you will know what I meant.

I know this is not a place to express any political views but just to let you know that I don't need to keep what I've been doing private ;), and your generosity plays a big part in changing what young people should learn about their history.
